### ISSUE TITLE
logger: track system metrics automatically

### ISSUE THREAD

**Issue by aguschin** [iterative/dvclive#81]
Right now dvclive doesn't track system metrics, to name a few, CPU, GPU, RAM utilisation. This is useful when comparing experiment results (e.g., I can get +1% accuracy, but what is the price in terms of time/resources?) and analysing how long it takes to run training on different GPUs (e.g., I can rent another GPU model and get 2x speedup?).

The usual practice is to log this somehow manually with https://github.com/giampaolo/psutil and analyse the results later, but because heavy experiments in ML require this quite often, IMO it makes sense to have this functionality out-of-the-box.

Also it would be great to have summary on these metrics in .json file produced in experiment to make quick decisions instead of diving too deep (e.g., the average CPU utilisation was 4 cores, my script doesn't utilise all 32 cores I have; the peak RAM utilisation was 8GB, that means I can rent a smaller server on aws to run this training; etc).

To name one example among ml tools, this is already tracked with W&B, see the bottom of dashboard: https://wandb.ai/stacey/estuary?workspace=user-lavanyashukla

This page states main metrics logged in W&B https://docs.wandb.ai/ref/app/features/system-metrics

If this will be useful, I could gather a list of metrics with notes describing the cases these metrics are helpful to a user in ml tasks.

**Comment by pared** [comment]
Well, currently summary is just a gathering of lastly logged metrics. It seems that those system metrics would provide a lot of information that some users might find useful. 
Some questions that needs answering before proceeding implementation:
1. How psutil gathers the data? If its from current point of time, we might get untrustworthy results (since logging usually happens in the end of epoch.
2. How to store the summary? Make `summary.json` just log all values? Or update the structure, to contain, for example "metrics", "system_performace_metrics" subdicts?

**Comment by dberenbaum** [comment]
>     1. How psutil gathers the data? If its from current point of time, we might get untrustworthy results (since logging usually happens in the end of epoch.

This is in the docs:

> wandb automatically logs system metrics every 2 seconds, averaged over a 30 second period.

**Comment by pared** [comment]
We could probably achieve that by creating `dvclive` context that would start "system observer" thread.

**Comment by daavoo** [comment]
This open source repo has a "hardware monitoring" functionality implemented with `psutil` (and `py3nvml` for GPUs):

https://github.com/labmlai/labml/blob/master/guides/hardware_monitoring.md

https://github.com/labmlai/labml/search?q=psutil

**Comment by daavoo** [comment]
I think this feature makes more sense to be implemented with a wider scope as part of DVC pipelines

**Comment by dberenbaum** [comment]
@daavoo Reopening to discuss. Does at least some of this info (CPU, GPU, RAM utilization) make sense in dvclive? For one thing, I think it's important to monitor those in realtime rather than only have a summary at the end.

**Comment by daavoo** [comment]
> Reopening to discuss. Does at least some of this info (CPU, GPU, RAM utilization) make sense in dvclive? For one thing, I think it's important to monitor those in realtime rather than only have a summary at the end.

Yes, I think a simple approach that collects the numbers and use `log_metric` makes sense to be implemented in DVCLive

**Comment by alex000kim** [comment]
Would adding methods like `log_cpu_utilization`, `log_ram_utilization` and `log_gpu_utilization` to `src/dvclive/live.py` work?
E.g. 
```
# src/dvclive/live.py
...
import psutil
import GPUtil # would introduce dependency on https://github.com/anderskm/gputil
... 

class Live:
...
  def log_cpu_utilization(self):
      cpu_percent = psutil.cpu_percent()
      self.log_metric("system/cpu_utilization", cpu_percent)
  
  def log_ram_utilization(self):
      memory_info = psutil.virtual_memory()
      self.log_metric("system/ram_utilization", memory_info.percent)
  
  def log_gpu_utilization(self):
          gpus = GPUtil.getGPUs()
          for i, gpu in enumerate(gpus):
              self.log_metric(f"system/gpu_{i}_utilization", gpu.load)
              self.log_metric(f"system/gpu_{i}_memory_utilization", gpu.memoryUtil)
              
  def next_step(self):
      if self._step is None:
          self._step = 0
      self.log_cpu_utilization()
      self.log_ram_utilization()
      self.log_gpu_utilization()
 ...
 
 ```
 I ran some tests with it and it _seems_ to be working.
 Though, I am not 100% sure the results are trustworthy considering @pared's comment:
 > How psutil gathers the data? If its from current point of time, we might get untrustworthy results (since logging usually happens in the end of epoch.

**Comment by shcheklein** [comment]
yep, I guess we just need to run this code on a separate thread and collect these metrics per second.

**Comment by daavoo** [comment]
> So it's probably better to call `log_*_utilization` inside framework-specific callbacks.

It can/should be inside `Live`, but as @shcheklein said, there needs to be a background process that collects the values. This is how other loggers do it:

https://github.com/wandb/wandb/blob/main/wandb/sdk/internal/system/assets/interfaces.py#L103
